API to correlate system time with external time sources

pabigot opened the pull request:
This PR has content I've been working on for about six months, to locally timestamp sensor observations with standard time rather than time-since boot.

It provides data structures to capture a timestamp in two different clocks, monitor the drift between those clocks, and using a base instant with estimated drift convert between the clocks.  This provides the core technology to convert between system uptime and an external continuous time scale like TAI (UTC without applying leap seconds).

A sample is provided that demonstrates using this technology to measure local clock error  between Nordic HFCLK and LFCLK.

There are some questions in the first comment that need to be resolved.  There are also some follow-on commits that will be submitted separately after this seems to be going forward:
* support for Bluetooth Current Time Service as a time source (working, needs cleanup and documentation)
* conversion of the similar functionality in the Maxim DS3231 driver to use the new API

pabigot committed:
lib/timeutil: add utilities to manage local/reference clock alignment

Provide data structures to capture a timestamp in two different
clocks, monitor the drift between those clocks, and using a base
instant with estimated drift convert between the clocks.

This provides the core technology to convert between system uptime and
an external continuous time scale like TAI (UTC without applying leap
seconds).

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
samples: boards: nrf: add clock_skew

Provide a demonstration of using the timeutil skew infrastructure to
measure the relative error of the two clock sources on Nordic boards.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
doc: add a section for timeutil APIs

Describe the role of these APIs, key concepts that they depend on, and
expose the low-level API.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Some notes on the design, with a few questions that should be answered before it's ready to merge:
* confirm name for `timeutil_realtime_*` functions
* is userspace support done right?

## What is this?

There's already a `timeutil_` namespace for civil-time operations. This PR extends it with data structures and functions that allow skewed translation between time scales.

It then adds API under `CONFIG_TIMEUTIL_REALTIME` that supports system-wide translation from the current local tick counter to the corresponding civil time, including accounting for local clock skew (which has significant impact on accuracy).

The API does not automatically do time maintenance.  Something has to invoke `timeutil_realtime_set()` at least once, and preferably periodically, as well as calculating and storing a skew to adjust for local clock error.

## Why isn't this clock_gettime(), gettimeofday(), or k_realtime_get()?

The first two belong to libc, and attempting to override them is mis-use.  It's possible it could be done for newlib using internal hooks, but that wouldn't help for toolchains that don't use newlib.  It seems better to provide an external

It's not `k_realtime_*()` mostly because it's entirely library code. I'm open to renaming it.

## Is `timeutil_realtime_*` userspace-safe?

It seems to be.  When I started I assumed I'd need system calls for this, and even made `timeutil_realtime_get_state()` one.  But the unit tests seem to have no problem with it.  None of the data involved is a kernel object (and don't see why it should be); the only thing that's questionable is use of `k_spinlock`.

Also this does use floating point to do the time translation (calculating skews in integer-only math isn't feasible due to the magnitude of the values and scale factors).  It does this math under lock.  IIRC FP in the kernel was once an issue for Linux.  Clearly it works fine in supervisor-only Zephyr, I'm just worried that something about userspace might introduce a problem (e.g. failing to restore FPU state).

henrikbrixandersen commented on doc/reference/misc/timeutil.rst:
```diff
@@ -0,0 +1,289 @@
+offsets separately.
+
+The inverse transformation is not standardized: APIs like ``mktime()``
+expect information about time zones.  Zephyr module provides this
```
Which zephyr module?

henrikbrixandersen commented on doc/reference/misc/timeutil.rst:
```diff
@@ -0,0 +1,289 @@
+======================
+
+The :option:`CONFIG_TIMEUTIL_REALTIME` flag enables API that allows
+zephyr to provide the current civil ("real") time that advances aligned
```
```suggestion
zephyr to provide the current civil ("wall clock") time that advances aligned
```

henrikbrixandersen commented on doc/reference/misc/timeutil.rst:
```diff
@@ -0,0 +1,289 @@
+
+The continuous analogue is `UNIX Leap Time <timeutil-RFC8536s2>`_ which
+is UNIX time plus all leap-second corrections added after the POSIX
+epoch (when TAI-UTC was 8 s).
```
I don't understand the "when TAI-UTY was 8 s" comment?

pabigot replied:
The Julian date for 1970-01-01 is [2440587.50000](https://www.aavso.org/cgi-bin/cal2jd.pl).

The [formula for TAI-UTC in effect](https://www.iers.org/IERS/EN/Science/Recommendations/resolutionB1.html) on that date is:

    TAI-UTC=   4.2131700 S + (MJD - 39126.) X 0.002592 S

where MJD = JD - 2400000.5.  This evaluates to 8.000082.

henrikbrixandersen commented on doc/reference/misc/timeutil.rst:
```diff
@@ -0,0 +1,289 @@
+   following TAI with an offset of -32.
+-  UNIX Leap Time: epoch of 1970-01-01T00:00:00Z, continuous following
+   TAI with an offset of -8.
+
```
I must have missed something in the text leading up to this, but I seek a clarification of those TAI offsets?

pabigot replied:
They're the TAI-UTC offset in effect at the corresponding epoch.  For the first two they can be determined using the leap  second table; the UNIX epoch one is calculated as shown above.

henrikbrixandersen commented on lib/os/Kconfig:
```diff
@@ -72,4 +72,12 @@ config PRINTK_SYNC
 	  interleaving with concurrent usage from another CPU or an
 	  preempting interrupt.
 
+config TIMEUTIL_REALTIME
+	bool "Support conversion between local and civil time"
+	help
+	  When true API is provided that enables applications to read
+	  the current time based on maintained offsets and skews
+	  between the local tick clock and an external time source,
+	  possibly an RTC.
+
```
What is the rationale behind the `CONFIG_TIMEUTIL_REALTIME` naming? "Real time" may be confusing here (Zephyr is a Real-Time Operating System).

Have you considered `CONFIG_TIMEUTIL_CIVIL_TIME`? Same goes for the naming of the API functions.

pabigot replied:
Yes, but the [consensus interpretation of "civil time"](https://en.wikipedia.org/wiki/Civil_time) is the local time, which is almost always UTC plus an offset (that generally changes twice a year).

I'm not entirely bound to that decision.  Options I've considered include:
* civil time (implies a local time zone time scale)
* calendar time (implies a representation, not a time scale)
* wall clock
* timescale
* real time

none of which are wonderful.  I went with `REALTIME` because it's what `CLOCK_REALTIME` on a POSIX system gives you, which is UTC, and I think that the most likely use of these functions will be to produce something that approximates or provides a capability similar to `CLOCK_REALTIME`.  So: `timeutil_realtime_`.

If "realtime" is too confusing because Zephyr is a "real-time operating system" (it's not), then we can come up with a consensus alternative. 

 After reviewing all this, I'm warming to "timescale", which doesn't imply anything that's false or misleading.

jfischer-no replied:
I think it is confusing because realtime term is subjective, from the designation alone, everyone could understand something different about what for example `timeutil_realtime_get()` does. Maybe CONFIG_TIMEUTIL_CTS/`timeutil_cts_get()`, for civil time scale (https://github.com/zephyrproject-rtos/zephyr/pull/28977/commits/144c887c94846ee92f1338dfeece6288d30c1ff3#diff-ca1507f5fcb1d2e792a8c80442ab64038f0e5d9939a38e6015b1f1b5d28a8431R34) ?

pabigot replied:
CTS in Bluetooth is "Current Time Service", which is one of the potential sources for "real" time data.  So I think that spelling would be confusing.

jfischer-no replied:
It is less confusing than REALTIME, at least for me :-)

henrikbrixandersen reviewed (commented):
Just reviewed the documentation for now. I will try to find the time to review the rest within the next week.

henrikbrixandersen commented:
@pabigot Would it be feasible to drop the `float skew` (I am a little worried about introducing floating point requirements in the kernel code) with a PPB representation? Like turning it upside down and do all internal calculations based on an integer PPB value? Do we loose too much precision?

pabigot commented:
> Would it be feasible to drop the `float skew`?

I tried that.  The dynamic range due to the need to translate between different frequencies causes problems even with 64-bit integer math.  There are overflows and underflows during the intermediate calculation that makes the skew estimate horribly unstable even when the clocks are stable.

This ties to #29569 in that while this is proposed to be in `lib/os` it's unclear whether it's "kernel" code.  It maybe used in but doesn't require supervisor privileges.  I'm thinking we need a "support library" category (#29876) also falls into that category.

jfischer-no reviewed (commented):
:+1:

pabigot commented on lib/os/timeutil.c:
```diff
@@ -68,3 +69,106 @@ time_t timeutil_timegm(const struct tm *tm)
+
+	if (((tsp->base.ref == 0) && (inst->ref > 0))
+	    || ((inst->ref > tsp->base.ref)
+		&& (inst->local > tsp->base.local))) {
```
?  It's consistent in this file.  IMO we don't have a consensus style guide to govern this, and I deeply dislike having to look to the end of the previous line to understand what the effect of the next operand will be.

jfischer-no replied:
nope, see L:117 https://github.com/zephyrproject-rtos/zephyr/pull/28977/commits/c266183d8c9898bb0b575c58ac9fba25e7400b16#diff-99955b2a278f37abdb13b510b059d51eebd4b73ad1f980b16e3a78a68e12dad7R117 which is typical style in Zephyr.

pabigot replied:
Ah; thanks, fixed.

pabigot commented on lib/os/timeutil.c:
```diff
@@ -68,3 +69,106 @@ time_t timeutil_timegm(const struct tm *tm)
+			tsp->latest = *inst;
+			rv = 1;
+		}
+	}
```
That falls under the consensus rule of a blank line after a closing brace; done.

pabigot commented:
I believe the major point of contention is the use of "realtime" in parts of the API. My reasoning is at https://github.com/zephyrproject-rtos/zephyr/pull/28977#discussion_r511636768, but the most compelling is that this infrastructure adds to Zephyr capability that matches what POSIX provides with `CLOCK_REALTIME`.  Had #19030 gotten any traction that would too likely have used `REALTIME`.

But as I noted, if there's a consensus that some other term is preferred it can be changed.

jukkar commented on subsys/shell/modules/date_service.c:
```diff
@@ -211,9 +246,62 @@ static int cmd_date_get(const struct shell *shell, size_t argc, char **argv)
+	struct timeutil_sync_state state;
+	int ret = timeutil_realtime_get_state(&cfg, &state);
+
+	shell_print(shell, "Date data from timeutil realtime API", ret);
```
looks like `%d` is missing

noelleclement commented on doc/reference/misc/timeutil.rst:
```diff
@@ -0,0 +1,289 @@
+
+Something in the application must invoke
+:c:func:`timeutil_realtime_set()` at least once with a reference
+timestamp and the :c:func:`k_uptime_ticks()` value corresponding to the
```
So essentially the 'seed' is created at the time `timeutil_realtime_set()` is called. Is there any additional information on when/where this should be called for optimal use of the API (I can imagine, the earlier the better).

noelleclement commented on doc/reference/misc/timeutil.rst:
```diff
@@ -0,0 +1,289 @@
+that the local clock is accurate.
+
+Applications that require precise time should use the other functions in
+this API to provide additional pairs of reference and local time,
```
So more than one external reference is possible? Or is this is just a 'trick' by doing the same actions multiple times with multiple external references? How would the differences between these transformations (/external source time) be handled?

noelleclement commented on doc/reference/misc/timeutil.rst:
```diff
@@ -0,0 +1,289 @@
+The API around :c:func:`timeutil_sync_state_update()` supports the first
+step of converting between continuous time scales.
+
+The second step requires external information including schedules of
```
So right now this API could still cause issues if the external time source wouldn't be monotonic (like most wallclock time sources, such as RTC, might not be)?

noelleclement replied:
Is there a way that adding such functionality would allow support of a discontinuous time source?

pabigot replied:
I don't believe it belongs in this API.  You'd use this API to get a timestamp in a continuous scale like TAI, then do adjustments using external tables and rules.

E.g. if local time is in TAI or GPS time (continuous monotonic), you'd need the leap second tables referenced elsewhere to figure out what the UTC offset is for that time.  Then you'd use the UTC time in conjunction with tables/rules for the local time zone to get the corresponding local time.  You can reverse that, with the problem that some local times are either ambiguous (1:30 am during a "fall back" adjustment) or non-existent (1:30 am during "spring forward").

If you're using this API with a UTC time source, which is fairly likely, you'd still need the leap second tables to deal with the discontinuity when a leap second occurred.  You can see that in the table up around like 215.

noelleclement replied:
(Quick add-on: the RTC in STM32 SoCs can be non-monotonic, if the steps are taken to implement daylight saving time, and these instructions can also be memorized (by bit-setting). However, I believe right now that is not part of the configuration, so won't be a problem right now, if the RTC counter driver would be used)

I was asking the question because I was wondering if that would make it possible for us to still use the API, but since we discussed this above, that's solved. 

If you do end up doing the long-duration data collection runs after the API has been merged, please do share! (I guess it could also be useful for the potential switching clock source in Nordic, right? Since that would change the base instant and skew.)

noelleclement reviewed (commented):
As discussed in the API meeting on 5-jan-2021, I reviewed the PR to see if it could meet our use-case.

Our use-case summarized:
- Power Management (for STM32 SoC without LPTIM), requires correct elapsed ticks announcement post-wake up (CPU shut off during sleep mode). 
- HFCLK (HSI) used for CPU & Zephyr system timer, LFCLK (RTC based on LSI/LSE) used for elapsed time reference during sleep mode and wake-up interrupt

- Possible applications of the API: 
  - Allow correct conversion between (LFCLK) RTC peripheral wall-clock time and ticks (required for post-wake up elapsed time announcement) 
  - Check clock skew before sleep (to ensure that the conversion post-sleep will have better results)

I hope that this would make it possible to *correctly* have a discontinuous clock source for the Zephyr system timer (since it shuts off during sleep), by having good conversion methods available for the elapsed time reference post-wake up. We think that this was one of the causes of a previous prototype not working correctly.
For this we would additionally need to allow use of the API in semi-low level (at least early in post-wake up/IRQ handler sequence), to ensure that the tick announcement doesn't interfere with the 'regular' tick announcements done by the system timer.

pabigot commented on samples/boards/nrf/clock_skew/src/main.c:
```diff
@@ -0,0 +1,234 @@
+#include <drivers/counter.h>
+#include <nrfx_clock.h>
+
+#define TIMER_NODE DT_NODELABEL(timer0)
```
This is covered in the README: 

> This sample uses the API for correlating time sources to measure the skew between HFCLK (used for the CPU) and LFCLK (used for system time).

It's pretty specific to Nordic, because it's an important issue when `k_busywait()` uses a different clock than `k_sleep()`.

Really all you need is "counter"s from two different sources.

noelleclement replied:
Okay, so then
-  `nordic_nrf_clock` is the chosen configured SoC system clock (through clock_control API?) == HFCLK (external oscilator) = used for CPU
- `timer0` is chosen through `CONFIG_COUNTER_TIMER0` (which seems to be part of a counter API driver, `counter_nrfx_timer`) == LFCLK = used for system timer

My guess is that with this set up, this sample also wouldn't be compatible with NRF's power management, right (if the HFCLK is shut off in the chosen sleep mode, at least)? Or is there a NRF-specific reason this wouldn't be an issue?

pabigot replied:
In Nordic neither HFCLK nor LFCLK are ever turned off.  They can theoretically change their source, which would make them monotonic and continuous but not *linear*, which is a pain, but the clock API isn't really set up for doing that so it's unlikely to happen.

noelleclement replied:
Ah alright. But it might still be problematic for other SoCs/Boards (like the STM32L1x SoCs). IMO it should be stated somewhere really clearly that this API is only compatible with power management if requirements xyz are met. 

From the top of my head, something along the lines of:
- if API is to be used throughout application: system timer source & external time source both aren't discontinuous (don't shut off during sleep)
- if either system timer source or external time source are discontinuous (due to shutting off during sleep), the API can be used only at boot (so before either sources has shut off) - (+ potentially the suggestions we discussed above, with the calibration skew calculation)

I have a feeling this doesn't cover everything, but just as an idea.

noelleclement commented on doc/reference/misc/timeutil.rst:
```diff
@@ -0,0 +1,289 @@
+* Translation between the continuous but inaccurate Zephyr time scale
+  and an accurate external stable time scale;
+* Translation between the stable time scale and the (possibly
+  discontinuous) civil time scale.
```
Do you mean that the civil time scale would be non-monotonic? That's how I interpreted your definition of the term, that you described [[here](https://github.com/zephyrproject-rtos/zephyr/issues/30590#issuecomment-744427593)]. 
Also, the stable time scale would/should be continuous(/monotonic) then?

pabigot replied:
monotonic is well-defined: each value observed is no less than the previous value.

continuous follows the math definition: no abrupt changes in value.  For time it's generally saying it's linear.

TAI is a timescale that is monotonic and continuous.  Each time mark (representation of an instant) is totally ordered and maps one-to-one to an instant in time.

UTC is a timescale that is monotonic and discontinuous.  Each time mark is totally ordered and increasing, but some may be skipped, e.g. 2021-12-31T23:59:59 would not exist if a negative leap second applied at the end of this year.  (Or 2021-12-31T23:59:59 might be followed by 2021-12-31T23:59:60 before 2022-01-01T00:00:00 if there was a positive leap second.)

European Summer Time and other daylight-aware time scales are civil timescales that are not monotonic, and therefore discontinuous.  "civil" because it's a time scale legally defined by civil authorities, like local governments.  Not monotonic because time goes backward by 1 hour in fall, causing the same time mark to apply to two instants at a one hour interval.

So civil timescales can be non-monotonic, but the time scales used by this API must be monotonic and linear, so any any discontinuities must be addressed by external adjustments, just would be if somebody wanted time on a device to be in the local timezone-aware timescale.

noelleclement replied:
Thanks, that clarifies it a lot. If you think it would fit in the documentation, I think this information could be very useful in there as well.

noelleclement commented on samples/boards/nrf/clock_skew/README.rst:
```diff
@@ -0,0 +1,116 @@
+********
+
+This sample uses the API for correlating time sources to measure the
+skew between HFCLK (used for the CPU) and LFCLK (used for system time).
```
Okay, makes sense, but this would inherently mean that measuring the skew with the discontinuous time source (for example, the HFCLK that shuts off during sleep), would only be able to be done before the time source 'becomes' discontinuous (e.g. when the SoC goes into sleep). 
Would it then still be possible / make sense to do the conversion (properly) later into the application? If not, that would kind of defeat the purpose for this whole API (for our use-case), right?

[I really hope you'll respond with 'oh this is not a problem, we can find a quick solution for this by adding xyz', because basically the whole use-case of the API for us is counting on this, although it makes a lot of sense that the time sources need to be continuous]

pabigot replied:
How this affects your application depends on details I don't know.  Let's say you wake up once an hour, run for two seconds, and believe your RTC chip exactly matches external time.  Then on startup you could use the API we've been discussing elsewhere to advance the Zephyr tick counter by the number of ticks in 59 minute 58 seconds (or, better, the difference between current time and last went-to-sleep time).  Unless you're measuring to microsecond resolution no skew that occurs within that two seconds is relevant, and you can just use 1 as the skew, and not bother calculating or storing it anywhere.

My expectation, from non-formal experiments with Nordic-based boards that use an external crystal for the local reference that need to be validated by long-running experiences, is that an approximate skew can be provided by default.  E.g. on most Nordic-based boards it's about 25 ppm.  It's possible that the skew can be board-dependent, and certain that it can be affected by temperature and passage of time, so in critical applications you would want to recalibrate it once in a while.  Recalibration does mean that the clocks must be running for the entire interval between the two captures.

In my case I deploy two sensors in a location at 2021-03-01T00:00:00Z and come back several months later.  Sensor A detected an event at a time that it converts to 2021-04-01T01:23:30Z.  At 25 ppm error over the 31 days in March the local time would be off by as much as 67 seconds, so if sensor B detected an event at 2021-04-01T01:22:25Z I don't know whether they were the same event.  If I can get the error down to 2 ppm I know it's not, because the error is less than 6 seconds over 31 days.

noelleclement replied:
Hmm interesting. [thinking out loud] then the plan for us could be to use the API to:
- per board run calibration tests over longer interval (without power management, so no sleep)
- run the skew calculation at boot / start application (before sleep has occured) as semi-calibration (since after sleep it wouldn't be possible)
- compare both?

We don't have a (time) critical application, but since the use-case is to use it as the (elapsed time during sleep) time reference for the system timer, I would say that would make it pretty critical. 

It could definitely be that above solution combined with the other API we've been discussing could actually cover everything.

noelleclement commented on samples/boards/nrf/clock_skew/src/main.c:
```diff
@@ -0,0 +1,234 @@
+		       false, false, false);
+}
+
+/* Convert HFCLK reference to microseconds. */
```
Great! Then if I may, I would suggest to include that in the API documentation, since I can imagine others might have this question too.

pabigot replied:
I'll see if I can find a place for it.

pabigot commented:
Rebased on master, and refactored so all the stuff that supports converting Zephyr uptime to realtime is removed.  It can be resubmited as a separate PR after the core functionality is merged.

pabigot commented on doc/reference/misc/timeutil.rst:
```diff
@@ -0,0 +1,240 @@
+representation of this time.
+
+Applications often need to correlate the Zephyr internal time with external
+time scales scales used in daily life, such as local time or Coordinated
```
Typo; fixed now.

jukkar reviewed (approved):
LGTM
